### PR TITLE
Add support for plans and terms

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/118/
Test PASSed.

**Comment by fabricematrat** [comment]
QA

python

from theblues.plans import Plans
from jujubundlelib import references
p = Plans("http://api.jujugui.org/omnibus/v2")
ref = references.Reference.from_string('cs:trusty/landscape-mock-0')
p.get_plans(ref)

from theblues.terms import Terms
t = Terms("http://api.jujugui.org/terms/v1")
t.get_terms("canonical")

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/119/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/120/
Test PASSed.

**Comment by kadams54** [comment]
When I run through the second bit, this is what I see:

~~~~
>>> t.get_terms("canonical")
theblues.errors.ServerError: unable to get terms for canonical: time data '2016-05-05T23:48:36.216Z' does not match format '%Y-%m-%dT%H:%M:%SZ'
~~~~

File: theblues/plans.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    log,
+    ServerError,
+)
+from theblues.utils import (
+    ensure_trailing_slash,
+    make_request,
+)
+
+Plan = namedtuple('Plan', 'url plan created_on description price')
+
```
Comment (jcsackett) [inline theblues/plans.py:16]: Looks like the second element of the named tuple should be several elements? e.g.

``` python
Plan = namedtuple('Plan', 'url',  'plan', 'created_on', 'description', 'price')
```

When you're adding below it looks like each of those is a separate element--am I misreading? Or am I misunderstanding named tuples?

Comment (jcsackett) [inline theblues/plans.py:16]: Ah, I am misunderstanding named tuples--but it looks like in the [namedtuple docs](https://docs.python.org/dev/library/collections.html#collections.namedtuple) it should be

``` python
Plan = namedtuple('Plan', ['url',  'plan', 'created_on', 'description', 'price'])
```

File: theblues/terms.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+from theblues.errors import (
+    log,
+    ServerError,
+)
+from theblues.utils import (
+    ensure_trailing_slash,
+    make_request,
+)
+
+Term = namedtuple('Term', 'name title revision created_on content')
```
Comment (jcsackett) [inline theblues/terms.py]: Same discussion as above on the Plan namedtuple.

**Comment by jcsackett** [comment]
@fabricematrat This looks pretty good to me--I have some concerns about the namedtuple declaration, but I'm not sure what you did was wrong. It just doesn't quite match the docs. It does look like your tests should fail if the declarations are wrong but we can see that they're passing. So I guess this is a "assuming what I wrote works, I would prefer it, but I can't be sure it's wrong so take it or leave it".

Thanks.

:+1: assuming the tuple issue isn't actually an issue.

File: theblues/plans.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from collections import namedtuple
```
Comment (frankban) [inline theblues/plans.py:1]: Please add license header.

File: theblues/plans.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+Plan = namedtuple('Plan', 'url plan created_on description price')
+
+
+class Plans(object):
+
+    def __init__(self, url):
+        """Initializer.
+
+        @param url The url to the Plan API.
```
Comment (frankban) [inline theblues/plans.py:23]: Should the URL include the revision? If so, please let's avoid it. The blues library should know what revision it supports, not users. Also, please document that.

File: theblues/plans.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        self.url = ensure_trailing_slash(url)
+
+    def get_plans(self, reference):
+        """ Get the plans for a given charm.
```
Comment (frankban) [inline theblues/plans.py]: Non required space in """ Get...

File: theblues/plans.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        @param the Reference to a charm
```
Comment (frankban) [inline theblues/plans.py]: Missing dot at the end of the sentence.

File: theblues/plans.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        @return a list of plans or an empty tuple if no plans.
```
Comment (frankban) [inline theblues/plans.py]: return a tuple of plans?

File: theblues/plans.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        @raise ServerError
+        """
+        json = make_request(
+            '{}charm?charm-url={}'.format(self.url,
+                                          'cs:' + reference.path()))
```
Comment (frankban) [inline theblues/plans.py:36]: IIRC the full path including cs schema can be generated by just doing `reference.id()`.

Comment (frankban) [inline theblues/plans.py:36]: Anyway, here it seems that the revision is dash separated. For instance it is "?charm-url=django-2", not "?charm-url=django/2". See http://api.jujugui.org/omnibus/v2/charm?charm-url=cs:trusty/landscape-mock-0.
So I guess it's like `'{}-{}.format(reference.with(revision=None).id(), reference.revision)` or similar.

File: theblues/terms.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Terms(object):
+
+    def __init__(self, url):
+        """Initializer.
+
+        @param url The url to the Terms Service API.
```
Comment (frankban) [inline theblues/terms.py:23]: Same as URL comment above.

Comment (frankban) [inline theblues/terms.py:23]: To be clear, let's define the revision being used here in the library and document that.

File: theblues/terms.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        """
+        self.url = ensure_trailing_slash(url)
+
+    def get_terms(self, name, revision=None):
+        """ Retrieve a specific term and condition.
+
+        @param name of the terms
+        @param revision of the terms
```
Comment (frankban) [inline theblues/terms.py]: Please mention that if revision is not provided the most recent revision is used.

File: theblues/terms.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        """Initializer.
+
+        @param url The url to the Terms Service API.
+        """
+        self.url = ensure_trailing_slash(url)
+
+    def get_terms(self, name, revision=None):
+        """ Retrieve a specific term and condition.
+
+        @param name of the terms
```
Comment (frankban) [inline theblues/terms.py]: Please complete the sentence.

File: theblues/terms.py (hunk 10 above)
Comment (frankban) [inline theblues/terms.py]: And please complete the sentence.

File: theblues/terms.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        @param revision of the terms
+        @return The list of terms.
+        @raise ServerError
+        """
+        url = '{}terms/{}'.format(self.url, name)
+        if revision:
+            url = '{}?revision={}'.format(url, revision)
+        json = make_request(url)
+        try:
+            # This is always a list of one element
```
Comment (frankban) [inline theblues/terms.py]: Guess what? Complete the sentence.

File: theblues/terms.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
+from collections import namedtuple
```
Comment (frankban) [inline theblues/terms.py:1]: License header.

File: theblues/tests/test_plan.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
+import datetime
```
Comment (frankban) [inline theblues/tests/test_plan.py:1]: License header.

File: theblues/tests/test_terms.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+import datetime
```
Comment (frankban) [inline theblues/tests/test_terms.py:1]: License header.

File: theblues/tests/test_plan.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+import json
+
+from mock import patch
+from unittest import TestCase
+
+from jujubundlelib import references
```
Comment (frankban) [inline theblues/tests/test_plan.py]: jujubundlelib should be in the second import group.

File: theblues/tests/test_plan.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+class TestPlans(TestCase):
+
+    def setUp(self):
+        self.plans = Plans('http://example.com')
+        self.ref = references.Reference.from_string(
+            'cs:trusty/landscape-mock-0')
+
+    def test_init(self):
+        self.assertEqual(self.plans.url, 'http://example.com/')
```
Comment (frankban) [inline theblues/tests/test_plan.py]: This should include the API version.

File: theblues/tests/test_terms.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+import json
+
+from mock import patch
+from unittest import TestCase
+
+
```
Comment (frankban) [inline theblues/tests/test_terms.py]: Remove this newline.

File: theblues/tests/test_terms.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
+import datetime
+import json
+
+from mock import patch
```
Comment (frankban) [inline theblues/tests/test_terms.py:3]: Please move mock in the second import group.

File: theblues/tests/test_terms.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+from theblues.errors import ServerError
+
+
+class TestTerms(TestCase):
+
+    def setUp(self):
+        self.terms = Terms('http://example.com')
+
+    def test_init(self):
+        self.assertEqual(self.terms.url, 'http://example.com/')
```
Comment (frankban) [inline theblues/tests/test_terms.py]: This should include the API version.

File: theblues/plans.py (hunk 1 above)
Comment (frankban) [inline theblues/plans.py:16]: It can be both ways.

**Comment by frankban** [comment]
:+1: with some changes, only minor stuff.
Thank you!

**Comment by kadams54** [comment]
QA is :ok:, thanks @fabricematrat!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/121/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/122/
Test PASSed.

**Comment by fabricematrat** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/theblues-lib-merge

**Pull request merged**